--- linux-3.0.35-Q7_IMX6-14.03.01.orig/drivers/mtd/devices/m25p80.c	2014-06-30 16:23:31.000000000 +0200
+++ linux-3.0.35-Q7_IMX6-14.03.01/drivers/mtd/devices/m25p80.c	2014-07-08 16:25:55.801248321 +0200
@@ -688,6 +688,11 @@
 	{ "mx25l12855e", INFO(0xc22618, 0, 64 * 1024, 256, 0) },
 	{ "mx25l25635e", INFO(0xc22019, 0, 64 * 1024, 512, 0) },
 	{ "mx25l25655e", INFO(0xc22619, 0, 64 * 1024, 512, 0) },
+	
+	/* Micron */
+	{ "n25q128a13",  INFO(0x20ba18, 0, 64 * 1024, 256, 0) },
+	{ "n25q128a11",  INFO(0x20bb18, 0, 64 * 1024, 256, 0) },
+	{ "n25q256a", INFO(0x20ba19, 0, 64 * 1024, 512, SECT_4K) },
 
 	/* Spansion -- single (large) sector size only, at least
 	 * for the chips listed here (without boot sectors).
